apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/SpannerTestDatabaseBase.cs

```diff
@@ -0,0 +1,224 @@
+
+    /// <summary>
+    /// Returns true if the database was created just for this test, or false if the database was an existing one
+    /// specified through an environment variable.
+    /// </summary>
+    public abstract bool Fresh { get; }
+
+    /// <summary>
+    /// Connection string including database.
+    /// </summary>
+    public string ConnectionString { get; }
+
+    /// <summary>
+    /// Connection string without the database.
+    /// </summary>
+    public string NoDbConnectionString { get; }
+
+    /// <summary>
+    /// The GCP Project ID to be used for the test database.
+    /// </summary>
+    public string ProjectId { get; }
+
+    /// <summary>
+    /// The fully-qualified database name.
+    /// </summary>
+    public DatabaseName DatabaseName { get; }
+
+    internal SpannerClientCreationOptions SpannerClientCreationOptions { get; }
+
+    protected SpannerTestDatabaseBase(string projectId, DatabaseDialect dialect)
+    {
+        TestLogger.Install();
+        ProjectId = projectId;
+        var builder = new SpannerConnectionStringBuilder
+        {
+            Host = SpannerHost,
+            DataSource = $"projects/{ProjectId}/instances/{SpannerInstance}",
+            EmulatorDetection = dialect == DatabaseDialect.Postgresql
+                ? EmulatorDetection.None
+                : EmulatorDetection.EmulatorOrProduction
+        };
+        if (SpannerPort != null)
+        {
+            builder.Port = int.Parse(SpannerPort);
+        }
+        NoDbConnectionString = builder.ConnectionString;
+        SpannerClientCreationOptions = new SpannerClientCreationOptions(builder);
+        var databaseBuilder = builder.WithDatabase(SpannerDatabase);
+        ConnectionString = databaseBuilder.ConnectionString;
+        DatabaseName = databaseBuilder.DatabaseName;
+
+        if (dialect != DatabaseDialect.Postgresql)
+        {
+            // We create database instance only on the emulator.
+            // In production, we expect the instance to already exist.
+            MaybeCreateInstanceOnEmulator(projectId);
+        }
+
+        if (Fresh)
+        {
+            CreateDatabase(dialect);
+            Logger.DefaultLogger.Debug($"Created database {SpannerDatabase} with {dialect} dialect.");
+        }
+        else
+        {
+            Logger.DefaultLogger.Debug($"Using existing database {SpannerDatabase}");
+        }
+    }
+
+    protected static string GetEnvironmentVariableOrDefault(string name, string defaultValue)
+    {
+        string value = Environment.GetEnvironmentVariable(name);
+        return string.IsNullOrEmpty(value) ? defaultValue : value;
+    }
+
+    protected virtual void CreateDatabase(DatabaseDialect dialect)
+    {
+        // PostgreSQL database can only be created by Admin APIs, which is not supported by Emulator for GSQL database creation.
+        // So, we need two different implementation for GSQL and PostgreSQL.
+        if (dialect == DatabaseDialect.Postgresql)
```

Comment:
Everything on this if should be the implementation of this method on the PS specific derived class.